### ISSUE TITLE
AdaptivePool does not match Torch outputs

### ISSUE THREAD

**Issue by Rocketknight1** [tensorflow/addons#2729]
Right now the adaptive pooling layers in tensorflow-addons do not match the behavior of their PyTorch equivalents. This is a problem as a number of transformer vision models have been released that use these layers, which makes these models impossible to port to TensorFlow. This issue has been raised in the past (#2322) but the implementation in that PR is slow because the output is constructed in an unvectorized way (e.g. [this block](https://github.com/tensorflow/addons/pull/2322/files#diff-4c343298abb27bb518ecef86bcad6e3e8645aa3ae9acb73f88f4941fa16421dfR279-R288))

We have written a much more performant version of the op which matches Torch's behavior exactly - you can see a gist [here](https://gist.github.com/Rocketknight1/efc47242914788def0144b341b1ad638). This gist could easily be extended to the other adaptive pool types.

We're happy to prepare a PR, but we have some questions!

1) These layers will have different behavior from the existing AdaptivePool layers in `tensorflow-addons`, because they now match torch exactly. Is this okay? Do we need to keep the old `tensorflow-addons` layer? Does anyone need that specific functionality?

2) These layers will probably be a little slower than the existing AdaptivePool layers in `tensorflow-addons`, because to mimic torch exactly we need to use two pools and a gather per axis. Is this okay?

**Relevant information**
- Are you willing to contribute it (yes/no): yes
- Are you willing to maintain it going forward? (yes/no): yes
- Is there a relevant academic paper? (if so, where): Yes. A number of major PyTorch vision models use this layer, such as [Swin Transformer](https://arxiv.org/abs/2103.14030).
- Does the relavent academic paper exceed 50 citations? (yes/no): Yes - 1755 citations for Swin Transformer alone.
- Is there already an implementation in another framework? (if so, where): Yes, [in PyTorch](https://pytorch.org/docs/stable/generated/torch.nn.AdaptiveAvgPool2d.html)
- Was it part of tf.contrib? (if so, where): No.

**Which API type would this fall under (layer, metric, optimizer, etc.)**
Layer

**Who will benefit with this feature?**
Anyone who wants to use modern vision networks in TF. The lack of this op is a major issue here at Hugging Face!

**Comment by bhack** [comment]
@Rocketknight1 Thanks, can propose a PR at https://github.com/keras-team/keras-cv/? As we are going to collect swin transformer there.

**Comment by Rocketknight1** [comment]
@bhack Sounds good!
